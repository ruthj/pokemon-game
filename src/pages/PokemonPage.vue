<template>
    <h1 v-if="!pokemon">Loading ...</h1>
    <div v-else>
      <h1>Who is this pokemon?</h1>
      <PokemonPicture
        :pokemonId="pokemon.id"
        :showPokemon="showPokemon" 
      />

      <PokemonOption 
        :pokemons="pokemonArr"
        @selection = "checkAnswer"
      />

      <div v-if="showAnswer">
        <h2>{{message}}</h2>
        <button @click="newGame">
          New Game
        </button>
      </div>
    </div>
    
</template>

<script>

import PokemonOption from '@/components/PokemonOption.vue'
import PokemonPicture from '@/components/PokemonPicture.vue'
import getPokemonOptions from '@/helpers/getPokemonOptions.js'

//console.log(getPokemonOptions())

export default {
  components: {PokemonOption, PokemonPicture},
  data(){
    return{
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: ''
    }
  },
  methods: {
    async mixPokemonArray(){
      this.pokemonArr = await getPokemonOptions()
      //console.log(this.pokemonArr)
      const pokemonPosition = Math.floor(Math.random() * 4)
      this.pokemon = this.pokemonArr[pokemonPosition]
    },
    checkAnswer(selectedId){
      this.showPokemon = true
      this.showAnswer = true

      if(selectedId === this.pokemon.id){
        this.message = `Correct!, ${this.pokemon.name}`
      }else {
        this.message = `Oops, it was, ${this.pokemon.name}`
      }

    },
    newGame(){
      //window.location.reload();
      this.showPokemon = false
      this.showAnswer = false
      this.pokemonArr = []
      this.pokemon = null
      this.mixPokemonArray()
    }
  },
  mounted(){
    //mounted, one time method
    this.mixPokemonArray()
  }
}
</script>
