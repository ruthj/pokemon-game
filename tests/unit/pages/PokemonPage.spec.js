import {shallowMount, mount} from '@vue/test-utils'
import PokemonPage from '../../../src/pages/PokemonPage'
import {pokemons} from '../mocks/pokemons.mock'

describe('PokemonPages Component', () => {
    
    let wrapper
    beforeEach(() => {
        wrapper = shallowMount(PokemonPage)
    })

    test('should match snapshot', () => {
        expect (wrapper.html()).toMatchSnapshot()
    })

    test('should call mixPokemon when mounted', () => {
        const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')
        const wrapper = shallowMount(PokemonPage)
        expect(mixPokemonArraySpy).toHaveBeenCalled()
    })

    test('should match snapshot when loaded pokemos', () => {
        //mounts all the components const wrapper = mount(PokemonPage, {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        expect(wrapper.html()).toMatchSnapshot()
        
    })

    test('should show components PokemonPicture and PokemonOptions', () => {
       
        // const wrapper = mount(PokemonPage, {
        //     data() {
        //         return {
        //             pokemonArr: pokemons,
        //             pokemon: pokemons[0],
        //             showPokemon: false,
        //             showAnswer: false,
        //             message: ''
        //         }
        //     }
        // })
    

        // const pokemonPicture = wrapper.findComponent('PokemonPicture')
        // const pokemonOption = wrapper.findComponent('PokemonOption')

        // expect(pokemonOption).toBeTruthy()
        // expect(pokemonPicture).toBeTruthy()


        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        const pokemonPicture = wrapper.find('pokemon-picture-stub')
        const pokemonOption = wrapper.find('pokemon-option-stub')

        expect(pokemonPicture.exists()).toBeTruthy()
        expect(pokemonOption.exists()).toBeTruthy()
        expect(pokemonPicture.attributes('pokemonid')).toBe('15')
        expect(pokemonOption.attributes('pokemons')).toBeTruthy()

        //expect(wrapper.find('pokemon-picture-stub').attributes('pokemonid')).toBe('15')
    })

    test('should test checkAnswer() ', async () => {
        
        const wrapper = shallowMount (PokemonPage, {
            data(){
                return{
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
              }
        })

        await wrapper.vm.checkAnswer(15)
        expect(wrapper.find('h2').exists()).toBeTruthy()
        expect(wrapper.vm.showPokemon).toBeTruthy()
        expect(wrapper.find('h2').text()).toBe(`Correct!, ${pokemons[0].name}`)

        await wrapper.vm.checkAnswer(107)
        expect(wrapper.vm.message).toBe(`Oops, it was, ${pokemons[0].name}`)
    })

})