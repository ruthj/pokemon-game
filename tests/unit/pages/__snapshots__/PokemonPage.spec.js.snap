// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPages Component should match snapshot 1`] = `<h1>Loading ...</h1>`;

exports[`PokemonPages Component should match snapshot when loaded pokemos 1`] = `
<div>
  <h1>Who is this pokemon?</h1>
  <pokemon-picture-stub pokemonid="15" showpokemon="false"></pokemon-picture-stub>
  <pokemon-option-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-option-stub>
  <!--v-if-->
</div>
`;
